### PR TITLE
[5.3] Allow collection partition by key and add preserving keys option

### PR DESCRIPTION
**Pull request by sileence**
Some small improvements to the partition method.

- You can partition a collection by key (typically a column with a boolean value like: published/draft, premium/free, featured/not featured, etc.). i.e. `$posts->partition('featured')`.

- I also added the preserve keys option similar to the `groupBy` method.

### CODE REVIEW

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -726,17 +726,20 @@ public function forPage($page, $perPage)
...
-     * Partition the collection into two array using the given callback.
+     * Partition the collection into two arrays using the given callback or key.
      *
-     * @param  callable  $callback
+     * @param  callable|string  $callback
+     * @param  bool  $preserveKeys
      * @return array
      */
-    public function partition(callable $callback)
+    public function partition($callback, $preserveKeys = false)
```
Comment (CristianLlanos) [inline src/Illuminate/Support/Collection.php]: I would suggest preserving keys by default. It'll be more natural IMO.

Comment (CristianLlanos) [inline src/Illuminate/Support/Collection.php]: Well, groupBy doesn't preverse keys by default 🤔

Comment (taylorotwell) [inline src/Illuminate/Support/Collection.php]: Collections operations generally should preserve keys by default and there should be no option not to as you can call array_values or ->values() if you don't want to.

**Comment by taylorotwell** [comment]
Remove preserveKeys option and make preserving keys the default please.

File: src/Illuminate/Support/Collection.php [hunk 2]
```diff
@@ -726,17 +726,20 @@ public function forPage($page, $perPage)
...
     {
         $partitions = [new static(), new static()];
```
Comment (JosephSilber) [inline src/Illuminate/Support/Collection.php]: You can destructure a collection with `list` no problem.

Comment (CristianLlanos) [inline src/Illuminate/Support/Collection.php]: Cool!

**Comment by sileence** [comment]
@taylorotwell I'm returning a collection because as @JosephSilber pointed out it works with `list` (the tests pass).

File: src/Illuminate/Support/Collection.php [hunk 3]
```diff
@@ -726,17 +726,19 @@ public function forPage($page, $perPage)
...
-     * @param  callable  $callback
-     * @return array
+     * @param  callable|string  $callback
+     * @return static
      */
-    public function partition(callable $callback)
+    public function partition($callback)
     {
-        $partitions = [new static(), new static()];
+        $partitions = new static([new static(), new static()]);
```
Comment (JosephSilber) [inline src/Illuminate/Support/Collection.php]: 1. We usually leave off the `()` for constructors with no arguments.
2. It'd be better to keep `$partitions` as a simple array, and only wrap it in a collection before returning. It's a minute performance improvement, but hey, why not.

Comment (sileence) [inline src/Illuminate/Support/Collection.php]: I'll update the PR

**Comment by taylorotwell** [comment]
Can we make this work with the higher order stuff?

**Comment by sileence** [comment]
@taylorotwell example?

**Comment by taylorotwell** [comment]
$collection->partition->free ... I'll look into it.

**Pull request merged**